modin/core/dataframe/pandas/dataframe/dataframe.py

```diff
@@ -3659,10 +3688,20 @@ def _compute_new_widths():
                 if new_partitions.size > 0:
                     for part in new_partitions.T[0]:
                         if part._length_cache is not None:
-                            new_lengths.append(part.length())
+                            new_lengths.append(
+                                part.length(materialize=self._materialize_in_loop)
+                            )
                         else:
                             new_lengths = None
                             break
+                    if not self._materialize_in_loop:
+                        filter_condition = (
+                            self._partition_mgr_cls._execution_wrapper.check_is_future
```

Comment:
I guess we could move `apply_function_on_selected_items` function into `_partition_mgr_cls`. By doing this, we improve the structure of the code, since the dataframe will not need to know that there is such an entity as `_execution_wrapper`.

Reply:
Makes sense. I am also getting the `self.materialize_func = self._partition_mgr_cls._execution_wrapper.materialize` which is applied on `apply_function_on_selected_items ` from _execution_wrapper, To hide _execution_wrapper from dataframe.  I have changed the `apply_function_on_selected_items`  to a new function 'materialize_futures'  function just materializes the elements in given list.